api/src/main/java/org/apache/iceberg/io/ClosingIterator.java

```diff
@@ -25,33 +25,30 @@
 
 public class ClosingIterator<T> implements Iterator<T> {
   private final CloseableIterator<T> iterator;
-  private boolean shouldClose = false;
 
   public ClosingIterator(CloseableIterator<T> iterator) {
     this.iterator = iterator;
-
   }
 
   @Override
   public boolean hasNext() {
     boolean hasNext = iterator.hasNext();
-    this.shouldClose = !hasNext;
+    if (!hasNext) {
+      close();
+    }
     return hasNext;
   }
 
   @Override
   public T next() {
-    T next = iterator.next();
+    return iterator.next();
+  }
 
-    if (shouldClose) {
-      // this will only be called once because iterator.next would throw NoSuchElementException
-      try {
-        iterator.close();
-      } catch (IOException e) {
-        throw new UncheckedIOException(e);
-      }
+  private void close() {
```

Comment:
you probably should make sure that `close()` is only called once

Reply:
> you probably should make sure that `close()` is only called once

`Closeable::close` is required to be idempotent - if calling close on a `Closeable` resource has any side-effects, we should fix that at the source.

Reply:
A cursory search of various closable iterators in the iceberg project indicates that not all are idempotent. I think we should raise a ticket to address the issue in the iceberg project at least. But in the meantime we can just check if this has already been closed.

Reply:
Done, and added a test.

Reply:
What is the behaviour if htis iterator is being used in parallel?

Reply:
The iterator isn't required to be thread-safe from what I can tell.